llm_jailbreaking_defense/defenses/selfreminder.py

```diff
@@ -0,0 +1,84 @@
+"""
+Yueqi Xie, Jingwei Yi, Jiawei Shao, Justin Curl, Lingjuan Lyu, Qifeng Chen, Xing Xie & Fangzhao Wu
+Defending ChatGPT against jailbreak attack via self-reminders
+Nature Machine Intelligence: https://www.nature.com/articles/s42256-023-00765-8
+"""
+
+from .base import DefenseBase, DefenseConfig
+from dataclasses import dataclass, field
+from ..models import TargetLM
+import warnings
+
+@dataclass
+class SelfReminderConfig(DefenseConfig):
+    prefix_only: int = field(default=False)
+    suffix_only: bool = field(default=False)
+    query_template: str = field(default='remind')
+    system_message_template: str = field(default='remind')
+    
+    def __post_init__(self):
+        self.defense_method = 'SelfReminder'
+
+    def load_from_args(self, args):
+        super().load_from_args(args)
+        self.prefix_only = args.prefix_only
+        self.suffix_only = args.suffix_only
+        self.query_template = args.query_template
+        self.system_message_template = args.system_message_template
+        
+class SelfReminderDefense(DefenseBase):
+    def __init__(self, config, preloaded_model, **kwargs):
+        super().__init__(config)
+        self.suffix_only = config.suffix_only
+        self.prefix_only = config.prefix_only
+        self.query_template = config.query_template
+        self.system_message_template = config.system_message_template
+
+    def defense(self, prompt, target_lm, response=None):
+
+        self_reminder_prompt = self._remind_suffix(prompt)
+        
+        target_lm.system_message_template = self._remind_prefix()
```

Comment:
There may be issues if one target LM is used for multiple defenses? Maybe you could provide `system_message_template` as an argument of `get_response`.

Reply:
Okay, I can do that. To be clear, you mean to change it to something like
 `def get_response(self, prompts_list, system_message_template=None, verbose=True, **kwargs)` ?
 I can do that and make the appropriate changes and testing.

Reply:
Yes